solidity/src/Polymer7683.sol

```diff
@@ -0,0 +1,208 @@
+        _handleRefundOrder(orderId);
+    }
+
+    // ============ Internal Functions ============
+
+    /**
+     * @notice Dispatches a settlement instruction by emitting a Filled event that will be proven on the origin chain
+     * @param _originDomain The domain to which the settlement message is sent
+     * @param _orderIds The IDs of the orders to settle
+     * @param _ordersFillerData The filler data for the orders
+     */
+    function _dispatchSettle(
+        uint32 _originDomain,
+        bytes32[] memory _orderIds,
+        bytes[] memory _ordersFillerData
+    ) internal override {
+        for (uint256 i = 0; i < _orderIds.length; i++) {
+            FilledOrder memory order = filledOrders[_orderIds[i]];
+            emit Filled(_orderIds[i], order.originData, order.fillerData);
+        }
+    }
+
+    /**
+     * @notice Dispatches a refund instruction by emitting a Refund event that will be proven on the origin chain
+     * @param _originDomain The domain to which the refund message is sent
+     * @param _orderIds The IDs of the orders to refund
+     */
+    function _dispatchRefund(
+        uint32 _originDomain,
+        bytes32[] memory _orderIds
+    ) internal override {
+        emit Refund(_orderIds);
+    }
+
+    /**
+     * @notice Retrieves the local domain identifier
+     * @return The local domain ID (chain ID)
+     */
+    function _localDomain() internal view override returns (uint32) {
+        return uint32(localChainId);
+    }
+
+    function _validateSettlementProof(
+        bytes calldata eventProof
+    ) private view returns (
+        bytes32 eventOrderId,
+        bytes memory originData,
+        bytes memory fillerData
+    ) {
+        (
+            uint32 provenChainId,
+            address actualEmitter,
+            ,  // topics
+            bytes memory data
+        ) = prover.validateEvent(eventProof);
+
+        // Decode settlement-specific data
+        (eventOrderId, originData, fillerData) = abi.decode(data, (bytes32, bytes, bytes));
+
+        // Validate chain ID with origin data
+        OrderData memory orderData = OrderEncoder.decode(originData);
+        if (provenChainId != orderData.destinationDomain) {
+            revert InvalidChainId();
+        }
+
+        // Verify destination contract is registered for the proven chain
+        address expectedEmitter = destinationContracts[provenChainId];
+        if (expectedEmitter == address(0)) revert UnregisteredDestinationChain();
+
+        // Validate emitter matches registered destination
+        if (actualEmitter != expectedEmitter) revert InvalidEmitter();
+    }
+
+    function _validateRefundProof(
+        bytes calldata eventProof
+    ) private view returns (bytes32[] memory eventOrderIds) {
+        (
+            uint32 provenChainId,
+            ,  // actualEmitter
+            ,  // topics
+            bytes memory data
+        ) = prover.validateEvent(eventProof);
+
```

Comment:
Here I think it should also validate that the `provenChainId` is the order's `destinationDomain"

Reply:
good call! I added validation logic to `_validateRefundProof`. here is the verification logic:
- we use the provenChainId to retrieve the expected contract address
- we compare the expected contract vs actual event emitter

I think it would be more secure if the Refund event contains a chainID that can be verified directly, but since this requires changes to the base events, I'd defer this to you guys.